concordium-consensus/src/Concordium/Scheduler/Environment.hs

```diff
@@ -488,40 +489,65 @@ class (StaticInformation m, ContractStateOperations m, MonadProtocolVersion m) =
   rejectingWith' !c reason = c >>= \case Right b -> return b
                                          Left a -> rejectTransaction (reason a)
 
+  -- |Add a contract upgrade to the 'ChangeSet'
+  addContractUpgrade :: ContractAddress
+                     -- ^The instance that should be upgraded.
+                     -> GSWasm.ModuleInterfaceA (InstrumentedModuleRef m GSWasm.V1)
+                     -- ^The new module to use for execution.
+                     -> m ()
+                                         
+
 -- |Index that keeps track of modifications of smart contracts inside a single
 -- transaction. This is used to cheaply detect whether a contract state has
 -- changed or not when a contract calls another.
 type ModificationIndex = Word
 
+-- |A pending update for v1 instances.
+data InstanceV1Update' mr = InstanceV1Update {
+    -- |The modification index.
+    index :: !ModificationIndex,
+    -- |Amount changed
+    amountChange :: !AmountDelta,
+    -- |Present if a state change has ocurred.
+    newState :: !(Maybe (UpdatableContractState GSWasm.V1)),
+    -- |Present if the contract has been upgraded.
+    -- Contract upgrades are only supported from PV 5 and onwards.    
+    newModule  :: !(Maybe (GSWasm.ModuleInterfaceA (mr GSWasm.V1)))
```

Comment:
We should probably record the new exposed receive methods here.

Reply:
Done